File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.kt
class Solution {
    fun countCharacters(words: Array<String>, chars: String): Int {
        var result = 0
        val freq = HashMap<Char, Int>()
        chars.forEach {
            freq[it] = if (freq.containsKey(it)) freq[it]!! + 1 else 1
        }

        words.forEach {
            var isGood = true
            val copied = freq.clone() as HashMap<Char, Int>
            it.forEach { it1 ->
                if (!copied.containsKey(it1) || copied[it1]!! == 0) {
                    isGood = false
                    return@forEach
                }

                copied[it1] = copied[it1]!! - 1
            }

            if (isGood) result += it.length
        }

        println(freq.toString())
        return result
    }
}

File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.swift
func getFrequency(_ word: String) -> Dictionary<Character, Int> {
    var result: Dictionary<Character, Int> = [:]
    
    for c in word {
        result[c] = (result[c] ?? 0) + 1
    }
    
    return result
}

func canForm(chars: Dictionary<Character, Int>, word: Dictionary<Character, Int>) -> Bool {
    var mutableChars = chars
    var isValid = true
    
    word.forEach({ key, value in
        if let vc = mutableChars[key] {
            if vc >= value {
                mutableChars[key] = vc - value
            } else {
                isValid = false
            }
        } else {
            isValid = false
        }
    })
    
    return isValid
}

class Solution {
    
    func countCharacters(_ words: [String], _ chars: String) -> Int {
        var result = 0
        let charsFreqs = getFrequency(chars)
        
        for word in words {
            let wordFreqs = getFrequency(word)
            
            if canForm(chars: charsFreqs, word: wordFreqs) {
                result += word.count
            }
        }
        
        return result
    }
    
}
